File: design-patterns/cpp/state/idle_state.cpp
#include "idle_state.h"
#include "vending_machine.h"
#include "item_selected_state.h"
#include <iostream>

void IdleState::selectItem(VendingMachine& context, const std::string& itemCode) {
    std::cout << "Item selected: " << itemCode << std::endl;
    context.setSelectedItem(itemCode);
    context.setState(std::make_unique<ItemSelectedState>());
}

void IdleState::insertCoin(VendingMachine& context, double amount) {
    std::cout << "Please select an item first." << std::endl;
}

void IdleState::dispenseItem(VendingMachine& context) {
    std::cout << "Please select an item first." << std::endl;
} 

File: design-patterns/golang/state/idle_state.go
package state

import "fmt"

// IdleState represents the state when the machine is waiting for item selection
type IdleState struct{}

func (s *IdleState) SelectItem(context *VendingMachine, item string) {
	fmt.Printf("Item '%s' selected. Please insert coins.\n", item)
	context.SetSelectedItem(item)
	context.SetState(&ItemSelectedState{})
}

func (s *IdleState) InsertCoin(context *VendingMachine, amount float64) {
	fmt.Println("Please select an item first.")
}

func (s *IdleState) DispenseItem(context *VendingMachine) {
	fmt.Println("Please select an item and insert coins first.")
}
